x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationDenialMessages.java

```diff
@@ -92,6 +103,60 @@ private static String authenticatedUserDescription(Authentication authentication
         return userText;
     }
 
+    static String rolesDescription(Subject subject, @Nullable AuthorizationInfo authorizationInfo) {
+        // We cannot print the roles if it's an API key or a service account (both do not have roles, but privileges)
+        if (subject.getType() != Subject.Type.USER) {
+            return "";
+        }
+
+        final StringBuilder sb = new StringBuilder();
+        final boolean hasDeclaredRoleNames = subject.getUser().roles().length > 0;
+        if (hasDeclaredRoleNames) {
+            sb.append(" with declared roles [").append(Strings.arrayToCommaDelimitedString(subject.getUser().roles())).append("]");
+        } else {
+            sb.append(" with no declared roles");
+        }
+        if (authorizationInfo == null) {
+            return sb.toString();
+        }
+
+        final Set<String> resolvedRoleNames;
+        final Role role = RBACEngine.maybeGetRBACEngineRole(authorizationInfo);
+        if (role == Role.EMPTY) {
+            resolvedRoleNames = Set.of();
```

Comment:
`Role.EMPTY` does not have a empty list of role names but `["__empty"]`. These code is to take care of this quirk, which I raised a separate discussion to address.